### PR TITLE
CI: upgrade test fixes

### PR DESCRIPTION
**Pull request by Jaakko-Os**
- this PR implements the upgrade tests run on top of ubuntu based environment

- the combined upgrade test, 1cp_1w_bootDiskImageANDK8sControllers_clusterLevel_upgrade, 
 commented out due to https://github.com/metal3-io/metal3-dev-env/issues/427 is run on top of centos environment. This is due to the need of running ironic in pods. Non-default CAPM3RELEASE and CAPIRELEASE environment variable handling is needed for the combined test. A separate PR will implement this part

### CODE REVIEW

File: scripts/feature_tests/setup_env.sh [hunk 1]
```diff
@@ -5,8 +5,9 @@ set -x
 M3PATH="$(dirname "$(readlink -f "${0}")")/../.."
 
 if [[ "${1}" == "ug" ]]; then
-  export CAPM3RELEASE="v0.3.2"
-  export CAPIRELEASE="v0.3.4"
+  # shellcheck disable=SC1091
+  # shellcheck source="$M3PATH/scripts/feature_tests/upgrade/upgrade_vars.sh"
```
Comment (fmuyassarov) [inline scripts/feature_tests/setup_env.sh:9]: Is this still needed?

**Comment by kashifest** [comment]
@Jaakko-Os I can see that you have taken changes from [PR#419 ](https://github.com/metal3-io/metal3-dev-env/pull/419). Should I close 419 then?

**Comment by Jaakko-Os** [comment]
> @Jaakko-Os I can see that you have taken changes from [PR#419 ](https://github.com/metal3-io/metal3-dev-env/pull/419). Should I close 419 then?

yes

**Comment by fmuyassarov** [comment]
> /test-upgrade-features

CI is going to fail  https://github.com/metal3-io/project-infra/pull/135 is not yet merged.

**Comment by kashifest** [comment]
> > /test-upgrade-features
> 
> CI is going to fail [metal3-io/project-infra#135](https://github.com/metal3-io/project-infra/pull/135) is not yet merged.

Whats keeping us from merging it?

**Comment by kashifest** [comment]
> > > /test-upgrade-features
> > 
> > 
> > CI is going to fail [metal3-io/project-infra#135](https://github.com/metal3-io/project-infra/pull/135) is not yet merged.
> 
> Whats keeping us from merging it?

Ow, @Jaakko-Os I think you need to push the branch again on https://github.com/metal3-io/project-infra/pull/135

**Comment by fmuyassarov** [comment]
> > > /test-upgrade-features
> > 
> > 
> > CI is going to fail [metal3-io/project-infra#135](https://github.com/metal3-io/project-infra/pull/135) is not yet merged.
> 
> Whats keeping us from merging it?

This PR is needed so that metal3-io/project-infra#135 is can pass the upgrade CI.

File: scripts/feature_tests/setup_env.sh (hunk 1 above)
Comment (Jaakko-Os) [reply scripts/feature_tests/setup_env.sh]: removed

File: scripts/feature_tests/setup_env.sh [hunk 2]
```diff
@@ -5,8 +5,9 @@ set -x
...
+  source "$M3PATH/scripts/feature_tests/upgrade/upgrade_vars.sh"
```
Comment (maelk) [inline scripts/feature_tests/setup_env.sh:10]: is this needed if you define the variables in the makefile ?

File: scripts/feature_tests/upgrade/upgrade.sh [hunk 3]
```diff
@@ -23,6 +23,10 @@ source "${METAL3_DEV_ENV_DIR}/lib/network.sh"
 # shellcheck disable=SC1090
 source "${METAL3_DEV_ENV_DIR}/lib/images.sh"
 
+# shellcheck disable=SC1091
+# shellcheck disable=SC1090
+source "${METAL3_DEV_ENV_DIR}/vm-setup/roles/v1aX_integration_test/files/manifests/clusterctl_env_${IMAGE_OS}.rc"
+
```
Comment (maelk) [inline scripts/feature_tests/upgrade/upgrade.sh:29]: what variables are you using from this file ? I am a bit concerned to rely on a file that was generated for a temporary use in ansible, and might be modified there, without realizing that it would impact this case, and breaking it. If you do not use many variables it might be better to define those variables here, or re-generate the file

File: scripts/feature_tests/upgrade/upgrade_vars.sh [hunk 4]
```diff
@@ -0,0 +1,6 @@
+#!/bin/bash
+
+set -x
+
+export CAPM3RELEASE="v0.3.2"
+export CAPIRELEASE="v0.3.4"
```
Comment (maelk) [inline scripts/feature_tests/upgrade/upgrade_vars.sh:6]: this does not look like it is used

Comment (maelk) [reply scripts/feature_tests/upgrade/upgrade_vars.sh:6]: I mean there is nothing sourcing this file as far as I can see

Comment (Jaakko-Os) [reply scripts/feature_tests/upgrade/upgrade_vars.sh:6]: true, the usage of non-default CAPM3RELEASE and CAPIRELEASE values sre commented out temporarily and this file is kept on purpose.

File: scripts/feature_tests/upgrade/upgrade.sh (hunk 3 above)
Comment (Jaakko-Os) [reply scripts/feature_tests/upgrade/upgrade.sh:29]: for the cluster provisioning:
API_ENDPOINT_HOST: 192.168.111.249
API_ENDPOINT_PORT: 6443
POD_CIDR: 192.168.0.0/18
SERVICE_CIDR: 10.96.0.0/12
and several other for control plane and worker provisioning

File: scripts/feature_tests/setup_env.sh (hunk 2 above)
Comment (Jaakko-Os) [reply scripts/feature_tests/setup_env.sh]: yes, based on local metal3-dev-env tests

File: scripts/feature_tests/upgrade/upgrade.sh (hunk 3 above)
Comment (Jaakko-Os) [reply scripts/feature_tests/upgrade/upgrade.sh]: removed

File: scripts/feature_tests/upgrade/1cp_1w_bootDiskImage_cluster_upgrade.sh [hunk 5]
```diff
@@ -33,9 +33,9 @@ wr_Metal3MachineTemplate_OUTPUT_FILE="/tmp/wr_new_image.yaml"
 CLUSTER_UID=$(kubectl get clusters -n metal3 test1 -o json | jq '.metadata.uid' |
```
Comment (furkatgofurov7) [inline scripts/feature_tests/upgrade/1cp_1w_bootDiskImage_cluster_upgrade.sh:33]:
```suggestion
CLUSTER_UID=$(kubectl get clusters -n "${NAMESPACE}" test1 -o json | jq '.metadata.uid' |
```

Comment (Jaakko-Os) [reply scripts/feature_tests/upgrade/1cp_1w_bootDiskImage_cluster_upgrade.sh]: done

File: vm-setup/roles/v1aX_integration_test/templates/clusterctl_env_template_ubuntu.rc [hunk 6]
```diff
@@ -57,8 +57,9 @@ export CTLPLANE_KUBEADM_EXTRA_CONFIG="
       - systemctl enable --now docker kubelet
       - if (curl -sk --max-time 10 https://{{ CLUSTER_APIENDPOINT_HOST }}:6443/healthz); then echo \"keepalived already running\";else systemctl start keepalived; fi
       - usermod -aG docker {{ IMAGE_USERNAME }}
-      - systemctl link /lib/systemd/system/monitor.keepalived.service 
-      - systemctl enable --now monitor.keepalived.service
+      - systemctl link /lib/systemd/system/monitor.keepalived.service
+      - systemctl enable monitor.keepalived.service
+      - systemctl start monitor.keepalived.service
```
Comment (kashifest) [inline vm-setup/roles/v1aX_integration_test/templates/clusterctl_env_template_ubuntu.rc:62]: Shouldn't it work with --now ?

Comment (Jaakko-Os) [reply vm-setup/roles/v1aX_integration_test/templates/clusterctl_env_template_ubuntu.rc:62]: it should, same outcome as enable + start

Comment (kashifest) [reply vm-setup/roles/v1aX_integration_test/templates/clusterctl_env_template_ubuntu.rc:62]: Then please do not change it

**Comment by furkatgofurov7** [comment]
@Jaakko-Os we should run integration tests, upgrade has passed!

File: scripts/feature_tests/upgrade/controlplane_upgrade/3cp_1w_k8sVer_bootDiskImage_scaleInWorker_upgrade.sh [hunk 7]
```diff
@@ -32,10 +32,11 @@ manage_node_taints "${CLUSTER_APIENDPOINT_IP}"
 
 scale_workers_to 0
 worker_has_correct_replicas 0
+expected_free_nodes 1
```
Comment (kashifest) [inline scripts/feature_tests/upgrade/controlplane_upgrade/3cp_1w_k8sVer_bootDiskImage_scaleInWorker_upgrade.sh:35]: Is this used anywhere?

File: vm-setup/roles/v1aX_integration_test/templates/clusterctl_env_template_ubuntu.rc (hunk 6 above)
Comment (kashifest) [reply vm-setup/roles/v1aX_integration_test/templates/clusterctl_env_template_ubuntu.rc:62]: I am still struggling to understand the need to do this change

**Review by kashifest — CHANGES_REQUESTED** [review]
Looks fine, but how many test cases this is now testing? Seems a couple of them are commented out. There seems to be a few arrangements/assumptions done here to pass the tests for examples the use CAPM3RELEASE or CAPIRELEASE, which versions are used? If the previous problem related to this are not solved and you are planning to solve them in a future PR, please update the PR description.

File: scripts/feature_tests/upgrade/controlplane_upgrade/3cp_1w_k8sVer_bootDiskImage_scaleInWorker_upgrade.sh (hunk 7 above)
Comment (Jaakko-Os) [reply scripts/feature_tests/upgrade/controlplane_upgrade/3cp_1w_k8sVer_bootDiskImage_scaleInWorker_upgrade.sh:35]: yes this checks and waits that one node is in 'ready' state before continuing

Comment (kashifest) [reply scripts/feature_tests/upgrade/controlplane_upgrade/3cp_1w_k8sVer_bootDiskImage_scaleInWorker_upgrade.sh:35]: oops, I missed that it is a function call, thanks

File: vm-setup/roles/v1aX_integration_test/templates/clusterctl_env_template_ubuntu.rc (hunk 6 above)
Comment (Jaakko-Os) [reply vm-setup/roles/v1aX_integration_test/templates/clusterctl_env_template_ubuntu.rc:62]: the outcome is the same. I would like to merge this and it can be changed later.

File: vm-setup/roles/v1aX_integration_test/templates/clusterctl_env_template_centos.rc [hunk 8]
```diff
@@ -49,7 +49,7 @@ export CTLPLANE_KUBEADM_EXTRA_CONFIG="
       - retrieve.configuration.files.sh \"https://raw.githubusercontent.com/kubernetes/release/{{ KUBERNETES_BINARIES_CONFIG_VERSION }}/cmd/kubepkg/templates/latest/deb/kubeadm/10-kubeadm.conf\" /etc/systemd/system/kubelet.service.d/10-kubeadm.conf
       - usermod -aG docker {{ IMAGE_USERNAME }}
       - systemctl enable --now docker keepalived kubelet
-      - systemctl link /etc/systemd/system/multi-user.target.wants/monitor.keepalived.service
+      - systemctl link /lib/systemd/system/monitor.keepalived.service
```
Comment (fmuyassarov) [inline vm-setup/roles/v1aX_integration_test/templates/clusterctl_env_template_centos.rc:52]: Sorry if I missed the point, but why is this change needed?

Comment (Jaakko-Os) [reply vm-setup/roles/v1aX_integration_test/templates/clusterctl_env_template_centos.rc:52]: the link creation failed when writing under /etc/. /lib/ is the "natural" place for service items (other service itmes are there too)

File: vm-setup/roles/v1aX_integration_test/templates/clusterctl_env_template_ubuntu.rc (hunk 6 above)
Comment (fmuyassarov) [inline vm-setup/roles/v1aX_integration_test/templates/clusterctl_env_template_ubuntu.rc:62]: What is the reason to separate --now into enable+start?

**Review by fmuyassarov — COMMENTED** [review]
Looks good. Just a final nit from my side.

File: vm-setup/roles/v1aX_integration_test/templates/clusterctl_env_template_ubuntu.rc (hunk 6 above)
Comment (Jaakko-Os) [reply vm-setup/roles/v1aX_integration_test/templates/clusterctl_env_template_ubuntu.rc:62]: Earlier when the monitor.keepalived.service custom service was implemented, both were tried out and enable+start was working better.
Many problems were encountered during the testing process so possible issues needed to be ruled out one by one.

**Review by fmuyassarov — APPROVED** [review]
